Subject: [PATCH] Port QPR2 SignedPackage certificate history

Port the QPR2 factory, bounded history parsing, equality/hash, accessors and
paired Java parcel representation. Keep the native SignedPackageParcel ABI
and Samsung's existing getPackageName handling, including literal "*", so
older allowlist consumers do not receive a new exception or broader matching.
Mirror the latest digest in the retained legacy private field. Preserve the
128-entry limit and rejection of empty/null digest entries. No signature,
permission, attestation or package-manager verification is bypassed.

diff --git a/smali/android/content/pm/SignedPackage.smali b/smali/android/content/pm/SignedPackage.smali
--- a/smali/android/content/pm/SignedPackage.smali
+++ b/smali/android/content/pm/SignedPackage.smali
@@ -22,11 +22,26 @@
     .end annotation
 .end field
 
+.field static final blacklist MAX_CERTIFICATE_HISTORY_SIZE:I = 0x80
+
+.field private static final blacklist WILDCARD_PACKAGE_NAME:Ljava/lang/String; = "*"
+
 
 # instance fields
 .field private final blacklist mCertificateDigest:[B
 
+.field private final blacklist mCertificateDigestHistory:Ljava/util/List;
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "Ljava/util/List<",
+            "[B>;"
+        }
+    .end annotation
+.end field
+
 .field private final blacklist mPackageName:Ljava/lang/String;
+
+.field private final blacklist mWildcardPackageName:Z
 
 
 # direct methods
@@ -43,43 +58,120 @@
 .end method
 
 .method public constructor blacklist <init>(Landroid/content/pm/SignedPackageParcel;)V
+    .locals 2
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    iget-object v1, p1, Landroid/content/pm/SignedPackageParcel;->packageName:Ljava/lang/String;
+
+    iput-object v1, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
+
+    const-string v0, "*"
+
+    invoke-virtual {v0, v1}, Ljava/lang/Object;->equals(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    iput-boolean v0, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
+
+    iget-object v0, p1, Landroid/content/pm/SignedPackageParcel;->certificateDigest:[B
+
+    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+
+    const/4 v1, 0x0
+
+    if-eqz v0, :unica_native_digest_set
+
+    invoke-static {v0}, Ljava/util/List;->of(Ljava/lang/Object;)Ljava/util/List;
+
+    move-result-object v1
+
+    :unica_native_digest_set
+    iput-object v1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    return-void
+.end method
+
+.method public constructor blacklist <init>(Landroid/os/Parcel;)V
+    .locals 2
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    invoke-virtual {p1}, Landroid/os/Parcel;->readString8()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-static {v0}, Ljava/util/Objects;->requireNonNull(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v0
+
+    check-cast v0, Ljava/lang/String;
+
+    const-string v1, "*"
+
+    invoke-virtual {v1, v0}, Ljava/lang/Object;->equals(Ljava/lang/Object;)Z
+
+    move-result v1
+
+    iput-boolean v1, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
+
+    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
+
+    invoke-static {p1}, Landroid/content/pm/SignedPackage;->readCertificateHistory(Landroid/os/Parcel;)Ljava/util/List;
+
+    move-result-object p1
+
+    iput-object p1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    const/4 v0, 0x0
+
+    if-eqz p1, :unica_current_digest_set
+
+    invoke-virtual {p0}, Landroid/content/pm/SignedPackage;->getCertificateDigest()[B
+
+    move-result-object v0
+
+    :unica_current_digest_set
+    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+
+    return-void
+.end method
+
+.method private constructor blacklist <init>(Ljava/lang/String;Ljava/util/List;)V
     .locals 1
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Ljava/lang/String;",
+            "Ljava/util/List<",
+            "[B>;)V"
+        }
+    .end annotation
 
     invoke-direct {p0}, Ljava/lang/Object;-><init>()V
 
-    iget-object v0, p1, Landroid/content/pm/SignedPackageParcel;->packageName:Ljava/lang/String;
-
-    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
-
-    iget-object p1, p1, Landroid/content/pm/SignedPackageParcel;->certificateDigest:[B
-
-    iput-object p1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    return-void
-.end method
-
-.method public constructor blacklist <init>(Landroid/os/Parcel;)V
-    .locals 1
-
-    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
-
-    invoke-virtual {p1}, Landroid/os/Parcel;->readString8()Ljava/lang/String;
+    const-string v0, "*"
+
+    invoke-virtual {v0, p1}, Ljava/lang/Object;->equals(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    iput-boolean v0, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
+
+    iput-object p1, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
+
+    iput-object p2, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    const/4 v0, 0x0
+
+    if-eqz p2, :unica_current_digest_set
+
+    invoke-virtual {p0}, Landroid/content/pm/SignedPackage;->getCertificateDigest()[B
 
     move-result-object v0
 
-    invoke-static {v0}, Ljava/util/Objects;->requireNonNull(Ljava/lang/Object;)Ljava/lang/Object;
-
-    move-result-object v0
-
-    check-cast v0, Ljava/lang/String;
-
-    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
-
-    invoke-virtual {p1}, Landroid/os/Parcel;->createByteArray()[B
-
-    move-result-object p1
-
-    iput-object p1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+    :unica_current_digest_set
+    iput-object v0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
 
     return-void
 .end method
@@ -87,13 +179,355 @@
 .method public constructor whitelist <init>(Ljava/lang/String;[B)V
     .locals 0
 
-    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
-
-    iput-object p1, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
-
-    iput-object p2, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+    if-eqz p2, :cond_0
+
+    invoke-static {p2}, Ljava/util/List;->of(Ljava/lang/Object;)Ljava/util/List;
+
+    move-result-object p2
+
+    goto :goto_0
+
+    :cond_0
+    const/4 p2, 0x0
+
+    :goto_0
+    invoke-direct {p0, p1, p2}, Landroid/content/pm/SignedPackage;-><init>(Ljava/lang/String;Ljava/util/List;)V
 
     return-void
+.end method
+
+.method private static blacklist byteArrayListEquals(Ljava/util/List;Ljava/util/List;)Z
+    .locals 5
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Ljava/util/List<",
+            "[B>;",
+            "Ljava/util/List<",
+            "[B>;)Z"
+        }
+    .end annotation
+
+    const/4 v0, 0x1
+
+    if-ne p0, p1, :cond_0
+
+    return v0
+
+    :cond_0
+    const/4 v1, 0x0
+
+    if-eqz p0, :cond_5
+
+    if-nez p1, :cond_1
+
+    goto :goto_1
+
+    :cond_1
+    invoke-interface {p0}, Ljava/util/List;->size()I
+
+    move-result v2
+
+    invoke-interface {p1}, Ljava/util/List;->size()I
+
+    move-result v3
+
+    if-eq v2, v3, :cond_2
+
+    return v1
+
+    :cond_2
+    move v2, v1
+
+    :goto_0
+    invoke-interface {p0}, Ljava/util/List;->size()I
+
+    move-result v3
+
+    if-ge v2, v3, :cond_4
+
+    invoke-interface {p0, v2}, Ljava/util/List;->get(I)Ljava/lang/Object;
+
+    move-result-object v3
+
+    check-cast v3, [B
+
+    invoke-interface {p1, v2}, Ljava/util/List;->get(I)Ljava/lang/Object;
+
+    move-result-object v4
+
+    check-cast v4, [B
+
+    invoke-static {v3, v4}, Ljava/util/Arrays;->equals([B[B)Z
+
+    move-result v3
+
+    if-nez v3, :cond_3
+
+    return v1
+
+    :cond_3
+    add-int/lit8 v2, v2, 0x1
+
+    goto :goto_0
+
+    :cond_4
+    return v0
+
+    :cond_5
+    :goto_1
+    return v1
+.end method
+
+.method private static blacklist byteArrayListHashCode(Ljava/util/List;)I
+    .locals 2
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Ljava/util/List<",
+            "[B>;)I"
+        }
+    .end annotation
+
+    const/4 v0, 0x0
+
+    if-nez p0, :cond_0
+
+    return v0
+
+    :cond_0
+    invoke-interface {p0}, Ljava/util/List;->iterator()Ljava/util/Iterator;
+
+    move-result-object p0
+
+    :goto_0
+    invoke-interface {p0}, Ljava/util/Iterator;->hasNext()Z
+
+    move-result v1
+
+    if-eqz v1, :cond_1
+
+    invoke-interface {p0}, Ljava/util/Iterator;->next()Ljava/lang/Object;
+
+    move-result-object v1
+
+    check-cast v1, [B
+
+    mul-int/lit8 v0, v0, 0x1f
+
+    invoke-static {v1}, Ljava/util/Arrays;->hashCode([B)I
+
+    move-result v1
+
+    add-int/2addr v0, v1
+
+    goto :goto_0
+
+    :cond_1
+    return v0
+.end method
+
+.method public static whitelist createWithCertificateDigestHistory(Ljava/lang/String;Ljava/util/List;)Landroid/content/pm/SignedPackage;
+    .locals 2
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Ljava/lang/String;",
+            "Ljava/util/List<",
+            "[B>;)",
+            "Landroid/content/pm/SignedPackage;"
+        }
+    .end annotation
+
+    if-eqz p1, :cond_3
+
+    invoke-interface {p1}, Ljava/util/List;->isEmpty()Z
+
+    move-result v0
+
+    if-nez v0, :cond_2
+
+    invoke-interface {p1}, Ljava/util/List;->size()I
+
+    move-result v0
+
+    const/16 v1, 0x80
+
+    if-gt v0, v1, :cond_1
+
+    invoke-interface {p1}, Ljava/util/List;->iterator()Ljava/util/Iterator;
+
+    move-result-object v0
+
+    :goto_0
+    invoke-interface {v0}, Ljava/util/Iterator;->hasNext()Z
+
+    move-result v1
+
+    if-eqz v1, :cond_3
+
+    invoke-interface {v0}, Ljava/util/Iterator;->next()Ljava/lang/Object;
+
+    move-result-object v1
+
+    check-cast v1, [B
+
+    if-eqz v1, :cond_0
+
+    goto :goto_0
+
+    :cond_0
+    new-instance p0, Ljava/lang/IllegalArgumentException;
+
+    const-string/jumbo p1, "certificateDigestHistory cannot contain null digests"
+
+    invoke-direct {p0, p1}, Ljava/lang/IllegalArgumentException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+
+    :cond_1
+    new-instance p0, Ljava/lang/IllegalArgumentException;
+
+    new-instance v0, Ljava/lang/StringBuilder;
+
+    const-string/jumbo v1, "certificateDigestHistory size exceeds limit: "
+
+    invoke-direct {v0, v1}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-interface {p1}, Ljava/util/List;->size()I
+
+    move-result p1
+
+    invoke-virtual {v0, p1}, Ljava/lang/StringBuilder;->append(I)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p1
+
+    invoke-direct {p0, p1}, Ljava/lang/IllegalArgumentException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+
+    :cond_2
+    new-instance p0, Ljava/lang/IllegalArgumentException;
+
+    const-string/jumbo p1, "certificateDigestHistory cannot be empty"
+
+    invoke-direct {p0, p1}, Ljava/lang/IllegalArgumentException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+
+    :cond_3
+    new-instance v0, Landroid/content/pm/SignedPackage;
+
+    invoke-direct {v0, p0, p1}, Landroid/content/pm/SignedPackage;-><init>(Ljava/lang/String;Ljava/util/List;)V
+
+    return-object v0
+.end method
+
+.method private static blacklist readCertificateHistory(Landroid/os/Parcel;)Ljava/util/List;
+    .locals 4
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Landroid/os/Parcel;",
+            ")",
+            "Ljava/util/List<",
+            "[B>;"
+        }
+    .end annotation
+
+    invoke-virtual {p0}, Landroid/os/Parcel;->readInt()I
+
+    move-result v0
+
+    if-gez v0, :cond_0
+
+    const/4 p0, 0x0
+
+    return-object p0
+
+    :cond_0
+    if-eqz v0, :cond_4
+
+    const/16 v1, 0x80
+
+    if-gt v0, v1, :cond_3
+
+    new-instance v1, Ljava/util/ArrayList;
+
+    invoke-direct {v1, v0}, Ljava/util/ArrayList;-><init>(I)V
+
+    const/4 v2, 0x0
+
+    :goto_0
+    if-ge v2, v0, :cond_2
+
+    invoke-virtual {p0}, Landroid/os/Parcel;->createByteArray()[B
+
+    move-result-object v3
+
+    if-eqz v3, :cond_1
+
+    invoke-interface {v1, v3}, Ljava/util/List;->add(Ljava/lang/Object;)Z
+
+    add-int/lit8 v2, v2, 0x1
+
+    goto :goto_0
+
+    :cond_1
+    new-instance p0, Landroid/os/BadParcelableException;
+
+    new-instance v0, Ljava/lang/StringBuilder;
+
+    const-string v1, "Certificate digest at index "
+
+    invoke-direct {v0, v1}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v0, v2}, Ljava/lang/StringBuilder;->append(I)Ljava/lang/StringBuilder;
+
+    const-string v1, " is null"
+
+    invoke-virtual {v0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-direct {p0, v0}, Landroid/os/BadParcelableException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+
+    :cond_2
+    return-object v1
+
+    :cond_3
+    new-instance p0, Landroid/os/BadParcelableException;
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Certificate history size exceeds limit: "
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, v0}, Ljava/lang/StringBuilder;->append(I)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v1}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-direct {p0, v0}, Landroid/os/BadParcelableException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+
+    :cond_4
+    new-instance p0, Landroid/os/BadParcelableException;
+
+    const-string v0, "Certificate history size cannot be 0"
+
+    invoke-direct {p0, v0}, Landroid/os/BadParcelableException;-><init>(Ljava/lang/String;)V
+
+    throw p0
 .end method
 
 
@@ -128,17 +562,17 @@
 
     iget-object v3, p1, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
 
-    invoke-virtual {v1, v3}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+    invoke-static {v1, v3}, Ljava/util/Objects;->equals(Ljava/lang/Object;Ljava/lang/Object;)Z
 
     move-result v1
 
     if-eqz v1, :cond_1
 
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    iget-object p1, p1, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    invoke-static {p0, p1}, Ljava/util/Arrays;->equals([B[B)Z
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    iget-object p1, p1, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    invoke-static {p0, p1}, Landroid/content/pm/SignedPackage;->byteArrayListEquals(Ljava/util/List;Ljava/util/List;)Z
 
     move-result p0
 
@@ -151,15 +585,47 @@
 .end method
 
 .method public whitelist getCertificateDigest()[B
+    .locals 1
+
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    if-eqz p0, :cond_0
+
+    invoke-interface {p0}, Ljava/util/List;->size()I
+
+    move-result v0
+
+    add-int/lit8 v0, v0, -0x1
+
+    invoke-interface {p0, v0}, Ljava/util/List;->get(I)Ljava/lang/Object;
+
+    move-result-object p0
+
+    check-cast p0, [B
+
+    return-object p0
+
+    :cond_0
+    new-instance p0, Ljava/lang/NullPointerException;
+
+    const-string/jumbo v0, "No certificate digest available"
+
+    invoke-direct {p0, v0}, Ljava/lang/NullPointerException;-><init>(Ljava/lang/String;)V
+
+    throw p0
+.end method
+
+.method public whitelist getCertificateDigestHistory()Ljava/util/List;
     .locals 0
-
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    invoke-static {p0}, Ljava/util/Objects;->requireNonNull(Ljava/lang/Object;)Ljava/lang/Object;
-
-    move-result-object p0
-
-    check-cast p0, [B
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "()",
+            "Ljava/util/List<",
+            "[B>;"
+        }
+    .end annotation
+
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
 
     return-object p0
 .end method
@@ -175,7 +641,7 @@
 .method public whitelist hasCertificateDigest()Z
     .locals 0
 
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
 
     if-eqz p0, :cond_0
 
@@ -194,9 +660,9 @@
 
     iget-object v0, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
 
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    invoke-static {p0}, Ljava/util/Arrays;->hashCode([B)I
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    invoke-static {p0}, Landroid/content/pm/SignedPackage;->byteArrayListHashCode(Ljava/util/List;)I
 
     move-result p0
 
@@ -211,6 +677,14 @@
     invoke-static {p0}, Ljava/util/Objects;->hash([Ljava/lang/Object;)I
 
     move-result p0
+
+    return p0
+.end method
+
+.method public whitelist isWildcardPackageName()Z
+    .locals 0
+
+    iget-boolean p0, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
 
     return p0
 .end method
@@ -222,23 +696,47 @@
 
     invoke-direct {v0}, Landroid/content/pm/SignedPackageParcel;-><init>()V
 
+    iget-boolean v1, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
+
+    if-eqz v1, :cond_0
+
+    const-string v1, "*"
+
+    goto :goto_0
+
+    :cond_0
     iget-object v1, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
 
+    :goto_0
     iput-object v1, v0, Landroid/content/pm/SignedPackageParcel;->packageName:Ljava/lang/String;
 
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
+    invoke-virtual {p0}, Landroid/content/pm/SignedPackage;->hasCertificateDigest()Z
+
+    move-result v1
+
+    if-eqz v1, :cond_1
+
+    invoke-virtual {p0}, Landroid/content/pm/SignedPackage;->getCertificateDigest()[B
+
+    move-result-object p0
+
+    goto :goto_1
+
+    :cond_1
+    const/4 p0, 0x0
+
+    :goto_1
     iput-object p0, v0, Landroid/content/pm/SignedPackageParcel;->certificateDigest:[B
 
     return-object v0
 .end method
 
 .method public whitelist test-api toString()Ljava/lang/String;
-    .locals 2
+    .locals 4
 
     new-instance v0, Ljava/lang/StringBuilder;
 
-    const-string v1, "SignedPackage{packageName="
+    const-string/jumbo v1, "SignedPackage{packageName="
 
     invoke-direct {v0, v1}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
 
@@ -246,32 +744,71 @@
 
     invoke-virtual {v0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    const-string v1, ", certificateDigest="
+    const-string v1, ", certificateDigestHistory="
 
     invoke-virtual {v0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    iget-object v1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
+    iget-object v1, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
 
     if-nez v1, :cond_0
 
     const-string/jumbo p0, "null"
 
+    invoke-virtual {v0, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    goto :goto_1
+
+    :cond_0
+    const-string/jumbo v1, "["
+
+    invoke-virtual {v0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    const/4 v1, 0x0
+
+    :goto_0
+    iget-object v2, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    invoke-interface {v2}, Ljava/util/List;->size()I
+
+    move-result v2
+
+    if-ge v1, v2, :cond_2
+
+    if-lez v1, :cond_1
+
+    const-string v2, ", "
+
+    invoke-virtual {v0, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    :cond_1
+    invoke-static {}, Ljava/util/HexFormat;->of()Ljava/util/HexFormat;
+
+    move-result-object v2
+
+    iget-object v3, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    invoke-interface {v3, v1}, Ljava/util/List;->get(I)Ljava/lang/Object;
+
+    move-result-object v3
+
+    check-cast v3, [B
+
+    invoke-virtual {v2, v3}, Ljava/util/HexFormat;->formatHex([B)Ljava/lang/String;
+
+    move-result-object v2
+
+    invoke-virtual {v0, v2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    add-int/lit8 v1, v1, 0x1
+
     goto :goto_0
 
-    :cond_0
-    invoke-static {}, Ljava/util/HexFormat;->of()Ljava/util/HexFormat;
-
-    move-result-object v1
-
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    invoke-virtual {v1, p0}, Ljava/util/HexFormat;->formatHex([B)Ljava/lang/String;
-
-    move-result-object p0
-
-    :goto_0
+    :cond_2
+    const-string/jumbo p0, "]"
+
     invoke-virtual {v0, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
+    :goto_1
     const-string/jumbo p0, "}"
 
     invoke-virtual {v0, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
@@ -286,13 +823,60 @@
 .method public whitelist writeToParcel(Landroid/os/Parcel;I)V
     .locals 0
 
+    iget-boolean p2, p0, Landroid/content/pm/SignedPackage;->mWildcardPackageName:Z
+
+    if-eqz p2, :cond_0
+
+    const-string p2, "*"
+
+    goto :goto_0
+
+    :cond_0
     iget-object p2, p0, Landroid/content/pm/SignedPackage;->mPackageName:Ljava/lang/String;
 
+    :goto_0
     invoke-virtual {p1, p2}, Landroid/os/Parcel;->writeString8(Ljava/lang/String;)V
 
-    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigest:[B
-
-    invoke-virtual {p1, p0}, Landroid/os/Parcel;->writeByteArray([B)V
+    iget-object p2, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    if-nez p2, :cond_1
+
+    const/4 p0, -0x1
+
+    invoke-virtual {p1, p0}, Landroid/os/Parcel;->writeInt(I)V
 
     return-void
-.end method
+
+    :cond_1
+    invoke-interface {p2}, Ljava/util/List;->size()I
+
+    move-result p2
+
+    invoke-virtual {p1, p2}, Landroid/os/Parcel;->writeInt(I)V
+
+    iget-object p0, p0, Landroid/content/pm/SignedPackage;->mCertificateDigestHistory:Ljava/util/List;
+
+    invoke-interface {p0}, Ljava/util/List;->iterator()Ljava/util/Iterator;
+
+    move-result-object p0
+
+    :goto_1
+    invoke-interface {p0}, Ljava/util/Iterator;->hasNext()Z
+
+    move-result p2
+
+    if-eqz p2, :cond_2
+
+    invoke-interface {p0}, Ljava/util/Iterator;->next()Ljava/lang/Object;
+
+    move-result-object p2
+
+    check-cast p2, [B
+
+    invoke-virtual {p1, p2}, Landroid/os/Parcel;->writeByteArray([B)V
+
+    goto :goto_1
+
+    :cond_2
+    return-void
+.end method
